core/services/workflows/engine.go

```diff
@@ -871,3 +925,48 @@ func executeSyncAndUnwrapSingleValue(ctx context.Context, cap capabilities.Callb
 
 	return l.Underlying[0], nil
 }
+
+// Logging keys
+const (
+	cIDKey = "capabilityID"
+	tIDKey = "triggerID"
+	wIDKey = "workflowID"
+	eIDKey = "executionID"
+	sIDKey = "stepID"
+	sRKey  = "stepRef"
+)
+
+type workflowError struct {
+	labels map[string]string
+	// err is the underlying error that caused this error
+	err error
+	// reason is a human-readable string that describes the error
+	reason string
+}
+
+func (e *workflowError) Error() string {
+	// declare in reverse order so that the error message is ordered correctly
+	orderedLabels := []string{sRKey, sIDKey, tIDKey, cIDKey, eIDKey, wIDKey}
+
+	errStr := ""
```

Comment:
A `strings.Builder` would be more efficient here

Reply:
I'd rather keep it dead simple here, its not a hot path. I'll keep this in mind if I do something similar for logging though, which should have a measurable impact. Thanks!